include/drjit-core/jit.h

```diff
@@ -2474,6 +2480,154 @@ extern JIT_EXPORT uint32_t jit_array_write(uint32_t target, uint32_t offset,
 
+/// Opaque data structure, storing the recording of a thread state
+struct Recording;
+
+/**
+ * \brief Start recording operations which can be replayed later without tracing
+ * operations.
+ *
+ * \param backend
+ *      The backend for which recording should be started.
+ *
+ * \param inputs
+ *      An array of input variable indices, which have to be
+ *      specified when starting the recording.
+ *      Different indices, representing other variables,
+ *      might be provided when replaying.
+ *      This function borrows the indices and does not
+ *      increment their refcount.
+ *
+ * \param n_inputs
+ *      The number of input variables for the recording
+ */
+extern JIT_EXPORT void
+jit_freeze_start(JitBackend backend, const uint32_t *inputs, uint32_t n_inputs);
+
+/**
+ * \brief Stop recording operations and return a struct containing the
+ * recording.
+ *
+ * The recording is returned as an opaque pointer and has to be destroyed
+ * afterwards using the \cjitc_freeze_destroy function.
+ *
+ * \param backend
+ *      The backend on which recording should be stoped.
+ *
+ *  \param outputs
+ *      An array of output variable indices.
+ *      When replaying these variables are returned from the replay function.
+ *
+ *  \param n_outputs
+ *      The number of output variables of the recording.
+ */
+extern JIT_EXPORT Recording *jit_freeze_stop(JitBackend backend,
+                                             const uint32_t *outputs,
+                                             uint32_t n_outputs);
+
+/**
+ * \brief Replay a recording with different inputs.
+ *
+ * Replaying a recording with different inputs results in different output
+ * variables.
+ * Their variable indices get written into the outputs array.
+ *
+ * \param recording
+ *      The recording to replay given different inputs.
+ *
+ * \param inputs
+ *      An array of input variable indices for replaying the recording.
+ *      The number of inputs taken from the array is equal to the number of
+ *      inputs supplied to the jit_start_record function.
+ *
+ * \param outpus
+ *      This array is filled with the output variable indices, created when
+ *      replaying the recording.
+ *      The size of the array has to be equal to the number of output variables
+ *      supplied to the jit_record_stop function.
+ */
+extern JIT_EXPORT void jit_freeze_replay(Recording *recording,
+                                         const uint32_t *inputs,
+                                         uint32_t *outputs);
+
+/**
+ * \brief Perform a dry run replay of a recording (if required), which does not
+ * perform any actual work.
+ *
+ * Dry running the recording, calculates the widths of all kernels, and might
+ * return 0 if the function has to be re-recorded.
+ * This is required to catch cases where the size of an input variable changes
+ * the compiled kernels, for example when performing scatter reductions in LLVM
```

Comment:
That's just an example I could come up with.
Currently we only return `false` in the dry run if we replay a `Expand` operation and the size changed (this is specific to LLVM).
We also return false for the `Compress` operation, since we cannot infer the size of the resulting variable in the dry run mode and have to retrace the kernel in that case.
Dry runs are only performed if a `Expand` operation has been detected, so this will only be of concern for the `LLVM` backend.

Should I add something like this to the comment as well?

Reply:
Yes, that would be great! Having a central list of troublesome or potentially troublesome operations somewhere is good. (The other one I remember was "Poke").

Reply:
Added the function. Is the pass by reference approach ok? Also added a comment, is that enough or should I expand more?